### PR TITLE
Adding advanced_pytorch example

### PR DESCRIPTION
**Pull request by cozek**
#### Reference Issues/PRs
https://github.com/adap/flower/pull/803

#### What does this implement/fix? Explain your changes.
Implements an advaned_pytorch example. Closely duplicates the advanced_tensorflow example.

#### Any other comments?

I added a `--toy` flag since my machine can't run the full 10 client simulation.
Please run the full simulation and let me know how it works. Happy to make any required changes.
I had a notebook for prototyping, removed it. Can add it in a later PR if needed.

### CODE REVIEW

**Comment by danieljanes** [comment]
Hi @cozek , thanks for this :) we'll review the PR shortly. In the meantime, could you add a reference to the new `advanced_pytorch` example in README.md (close to where the `advanced_tensorflow` example is listed)?

**Comment by pedropgusmao** [comment]
Hi @cozek  the PR is looking great. Could you just set `DEVICE=cpu` in :
https://github.com/cozek/flower/blob/d36170c6785ca08716dd6ee2b6a1f184d6015dfb/examples/advanced_pytorch/utils.py#L9
It is likely that we run out of CUDA memory if not.

**Comment by pedropgusmao** [comment]
Another thing, each client appears to be downloading the CIFAR10 dataset. This creates problems when launching multiple clients. I suggest moving the download functionality to ./run.sh and simply loading the file later.

**Comment by cozek** [comment]
> Another thing, each client appears to be downloading the CIFAR10 dataset. This creates problems when launching multiple clients. I suggest moving the download functionality to ./run.sh and simply loading the file later.

Done.

**Comment by pedropgusmao** [comment]
@cozek , it's looking good. 
I just ran a few tests and apparently running in CPU takes about 25min per round... a bit too much :).
To solve this CUDA_OUT_OF_MEMORY vs CPU long training times, I suggest the following:

-  Store the model as CPU type inside the clients and just use it with cuda during training (remembering to store it back as CPU). 
- Reduce the number o participating clients to 2 per round.  
- Reduce `bach_size` to 16.

You can then keep an eye on your GPU memory usage which hopefully won't be too much.

**Comment by cozek** [comment]
> @cozek , it's looking good. I just ran a few tests and apparently running in CPU takes about 25min per round... a bit too much :). To solve this CUDA_OUT_OF_MEMORY vs CPU long training times, I suggest the following:
> 
> * Store the model as CPU type inside the clients and just use it with cuda during training (remembering to store it back as CPU).
> * Reduce the number o participating clients to 2 per round.
> * Reduce `bach_size` to 16.
> 
> You can then keep an eye on your GPU memory usage which hopefully won't be too much.

Done.

**Comment by pedropgusmao** [comment]
```
Traceback (most recent call last):
  File "server.py", line 108, in <module>
    main()
  File "server.py", line 104, in main
    fl.server.start_server("0.0.0.0:8080", config={"num_rounds": 4}, strategy=strategy)
  File "/home/pedro/repos/advanced_pytorch/src/py/flwr/server/app.py", line 114, in start_server
    force_final_distributed_eval=force_final_distributed_eval,
  File "/home/pedro/repos/advanced_pytorch/src/py/flwr/server/app.py", line 148, in _fl
    hist = server.fit(num_rounds=config["num_rounds"])
  File "/home/pedro/repos/advanced_pytorch/src/py/flwr/server/server.py", line 87, in fit
    res = self.strategy.evaluate(parameters=self.parameters)
  File "/home/pedro/repos/advanced_pytorch/src/py/flwr/server/strategy/fedavg.py", line 178, in evaluate
    eval_res = self.eval_fn(weights)
  File "server.py", line 60, in evaluate
    loss, accuracy = utils.test(model, valset)
  File "/home/pedro/repos/advanced_pytorch/examples/advanced_pytorch/utils.py", line 86, in test
    images, labels = images.to(DEVICE), labels.to(DEVICE)
AttributeError: 'int' object has no attribute 'to'
```

I'm getting the error above. Maybe you are returning the labels as an integer (original CIFAR). Instead, it must be at `Long Tensor` (int64) . I'm not sure if `range` is actually generating a list or if it must  be inside `list().`
Also, `cuda_device` in `train()` is not used.

**Comment by cozek** [comment]
> ```
> Traceback (most recent call last):
>   File "server.py", line 108, in <module>
>     main()
>   File "server.py", line 104, in main
>     fl.server.start_server("0.0.0.0:8080", config={"num_rounds": 4}, strategy=strategy)
>   File "/home/pedro/repos/advanced_pytorch/src/py/flwr/server/app.py", line 114, in start_server
>     force_final_distributed_eval=force_final_distributed_eval,
>   File "/home/pedro/repos/advanced_pytorch/src/py/flwr/server/app.py", line 148, in _fl
>     hist = server.fit(num_rounds=config["num_rounds"])
>   File "/home/pedro/repos/advanced_pytorch/src/py/flwr/server/server.py", line 87, in fit
>     res = self.strategy.evaluate(parameters=self.parameters)
>   File "/home/pedro/repos/advanced_pytorch/src/py/flwr/server/strategy/fedavg.py", line 178, in evaluate
>     eval_res = self.eval_fn(weights)
>   File "server.py", line 60, in evaluate
>     loss, accuracy = utils.test(model, valset)
>   File "/home/pedro/repos/advanced_pytorch/examples/advanced_pytorch/utils.py", line 86, in test
>     images, labels = images.to(DEVICE), labels.to(DEVICE)
> AttributeError: 'int' object has no attribute 'to'
> ```
> 
> I'm getting the error above. Maybe you are returning the labels as an integer (original CIFAR). Instead, it must be at `Long Tensor` (int64) . Also, `cuda_device` in `train()` is not used.

Yikes! I get back to you after I debug this.

**Comment by cozek** [comment]
> ```
> Traceback (most recent call last):
>   File "server.py", line 108, in <module>
>     main()
>   File "server.py", line 104, in main
>     fl.server.start_server("0.0.0.0:8080", config={"num_rounds": 4}, strategy=strategy)
>   File "/home/pedro/repos/advanced_pytorch/src/py/flwr/server/app.py", line 114, in start_server
>     force_final_distributed_eval=force_final_distributed_eval,
>   File "/home/pedro/repos/advanced_pytorch/src/py/flwr/server/app.py", line 148, in _fl
>     hist = server.fit(num_rounds=config["num_rounds"])
>   File "/home/pedro/repos/advanced_pytorch/src/py/flwr/server/server.py", line 87, in fit
>     res = self.strategy.evaluate(parameters=self.parameters)
>   File "/home/pedro/repos/advanced_pytorch/src/py/flwr/server/strategy/fedavg.py", line 178, in evaluate
>     eval_res = self.eval_fn(weights)
>   File "server.py", line 60, in evaluate
>     loss, accuracy = utils.test(model, valset)
>   File "/home/pedro/repos/advanced_pytorch/examples/advanced_pytorch/utils.py", line 86, in test
>     images, labels = images.to(DEVICE), labels.to(DEVICE)
> AttributeError: 'int' object has no attribute 'to'
> ```
> 
> I'm getting the error above. Maybe you are returning the labels as an integer (original CIFAR). Instead, it must be at `Long Tensor` (int64) . Also, `cuda_device` in `train()` is not used.

I fixed the issue. But I am unable to test the GPU utilization. I ran it in  kaggle notebook GPU using the changes you mentioned and still ran out of memory. I do not have access to a better GPU atm. Would appreciate Adap team's help in this regard.

**Comment by cozek** [comment]
@pedropgusmao @danieljanes ping.

**Comment by pedropgusmao** [comment]
Hello @cozek ,
Thanks for this.
This is what I am getting now:
```
INFO flower 2022-04-13 07:55:30,542 | server.py:127 | initial parameters (loss, other metrics): 0.14576197609901428, {'accuracy': 0.0752}
INFO flower 2022-04-13 07:55:30,542 | server.py:133 | FL starting
DEBUG flower 2022-04-13 07:55:30,542 | server.py:255 | fit_round: strategy sampled 2 clients (out of 10)
Starting training...
Starting training...
DEBUG flower 2022-04-13 07:55:36,192 | connection.py:68 | Insecure gRPC channel closed
Traceback (most recent call last):
  File "client.py", line 130, in <module>
    main()
  File "client.py", line 126, in main
    fl.client.start_numpy_client("0.0.0.0:8080", client=client)
  File "/home/pedro/.pyenv/versions/pytorch_example/lib/python3.7/site-packages/flwr/client/app.py", line 115, in start_numpy_client
    grpc_max_message_length=grpc_max_message_length,
  File "/home/pedro/.pyenv/versions/pytorch_example/lib/python3.7/site-packages/flwr/client/app.py", line 66, in start_client
    client, server_message
  File "/home/pedro/.pyenv/versions/pytorch_example/lib/python3.7/site-packages/flwr/client/grpc_client/message_handler.py", line 40, in handle
    return _fit(client, server_msg.fit_ins), 0, True
  File "/home/pedro/.pyenv/versions/pytorch_example/lib/python3.7/site-packages/flwr/client/grpc_client/message_handler.py", line 57, in _fit
    fit_res = client.fit(fit_ins)
  File "/home/pedro/.pyenv/versions/pytorch_example/lib/python3.7/site-packages/flwr/client/numpy_client.py", line 192, in fit
    results = self.numpy_client.fit(parameters, ins.config)
  File "client.py", line 55, in fit
    results = utils.train(self.model, trainLoader, valLoader, epochs)
  File "/home/pedro/repos/pytorch_example/examples/advanced_pytorch/utils.py", line 59, in train
    loss = criterion(net(images), labels)
  File "/home/pedro/.pyenv/versions/pytorch_example/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1051, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/pedro/.cache/torch/hub/NVIDIA_DeepLearningExamples_torchhub/PyTorch/Classification/ConvNets/image_classification/models/efficientnet.py", line 233, in forward
    x = self.layers(x)
  File "/home/pedro/.pyenv/versions/pytorch_example/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1051, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/pedro/.pyenv/versions/pytorch_example/lib/python3.7/site-packages/torch/nn/modules/container.py", line 139, in forward
    input = module(input)
  File "/home/pedro/.pyenv/versions/pytorch_example/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1051, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/pedro/.pyenv/versions/pytorch_example/lib/python3.7/site-packages/torch/nn/modules/container.py", line 139, in forward
    input = module(input)
  File "/home/pedro/.pyenv/versions/pytorch_example/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1051, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/pedro/.cache/torch/hub/NVIDIA_DeepLearningExamples_torchhub/PyTorch/Classification/ConvNets/image_classification/models/efficientnet.py", line 443, in forward
    self.se(self.depsep(x if self.expand is None else self.expand(x)))
  File "/home/pedro/.pyenv/versions/pytorch_example/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1051, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/pedro/.cache/torch/hub/NVIDIA_DeepLearningExamples_torchhub/PyTorch/Classification/ConvNets/image_classification/models/common.py", line 248, in forward
    out = self._attention(x)
  File "/home/pedro/.cache/torch/hub/NVIDIA_DeepLearningExamples_torchhub/PyTorch/Classification/ConvNets/image_classification/models/common.py", line 159, in _attention
    out = self.squeeze(out)
  File "/home/pedro/.pyenv/versions/pytorch_example/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1051, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/pedro/.pyenv/versions/pytorch_example/lib/python3.7/site-packages/torch/nn/modules/linear.py", line 96, in forward
    return F.linear(input, self.weight, self.bias)
  File "/home/pedro/.pyenv/versions/pytorch_example/lib/python3.7/site-packages/torch/nn/functional.py", line 1847, in linear
    return torch._C._nn.linear(input, weight, bias)
RuntimeError: CUDA error: CUBLAS_STATUS_ALLOC_FAILED when calling `cublasCreate(handle)`
Starting evalutation...
Starting evalutation...
DEBUG flower 2022-04-13 07:58:32,681 | server.py:264 | fit_round received 1 results and 1 failures
Starting evalutation...
INFO flower 2022-04-13 07:59:19,023 | server.py:154 | fit progress: (1, 0.19433022606372832, {'accuracy': 0.1756}, 228.48129541199887)
INFO flower 2022-04-13 07:59:19,023 | server.py:199 | evaluate_round: no clients selected, cancel

```

**Comment by danieljanes** [comment]
@pedropgusmao @cozek one idea: should we merge a version of this that fixes the torch device to `cpu`?

I think this example will provide a lot of value to the community, many users have asked for an advanced PyTorch example and I'd like to merge this as soon as possible. We could then follow up with a small patch PR that enables GPU execution.

**Comment by cozek** [comment]
> @pedropgusmao @cozek one idea: should we merge a version of this that fixes the torch device to `cpu`?
> 
> I think this example will provide a lot of value to the community, many users have asked for an advanced PyTorch example and I'd like to merge this as soon as possible. We could then follow up with a small patch PR that enables GPU execution.

Fine with me.

**Comment by cozek** [comment]
> @danieljanes @pedropgusmao on a side note, I would be really interested to know what changes are required to convert this shell script example into GPU compatible (since other simulation pytorch example abstracts a lot ray based stuff under the hood).

This script is GPU compatible, at least it I wrote it that way. Would appreciate it if you could test the script and see if it throws the same bug and if yes, how to go about fixing that. :)

**Comment by danieljanes** [comment]
@pedropgusmao thoughts?

**Comment by pedropgusmao** [comment]
I'm still getting out of memory from CUDA.
```
RuntimeError: CUDA out of memory. Tried to allocate 12.00 MiB (GPU 0; 5.79 GiB total capacity; 882.10 MiB already allocated; 45.88 MiB free; 922.00 MiB reserved in total by PyTorch)
```
Then I believe clients die and the server keeps waiting for them to continue. 
I think the model is still being stored in the client. Maybe if we just created it inside a `set_parameters` (`set_parameters` would then need to return a model), then we would not need to store the model itself. 
If that doesn't work we might need to move to CPU at least as a default option, with a possible flag for choosing GPU.

File: examples/advanced_pytorch/client.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+class CifarClient(fl.client.NumPyClient):
+    def __init__(
+        self,
+        model: torch.nn.Module,
+        trainset: torchvision.datasets,
+        testset: torchvision.datasets,
+        validation_split: int = 0.1,
+    ):
+        self.model = model
```
[the remark this answers is not in the document]
Comment (pedropgusmao) [reply examples/advanced_pytorch/client.py:21]: Maybe remove this here not to store the model.

File: examples/advanced_pytorch/client.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+from collections import OrderedDict
+import warnings
+
+warnings.filterwarnings("ignore")
+
+
+class CifarClient(fl.client.NumPyClient):
+    def __init__(
+        self,
+        model: torch.nn.Module,
```
Comment (pedropgusmao) [inline examples/advanced_pytorch/client.py:16]:
```suggestion
```
model can be created when calling `fit`

File: examples/advanced_pytorch/client.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        self.validation_split = validation_split
+
+    def get_parameters(self):
+        """Get parameters of the local model."""
+        raise Exception("Not implemented (server-side parameter initialization)")
+
+    def set_parameters(self, parameters):
+        params_dict = zip(self.model.state_dict().keys(), parameters)
+        state_dict = OrderedDict({k: torch.tensor(v) for k, v in params_dict})
+        self.model.load_state_dict(state_dict, strict=True)
```
Comment (pedropgusmao) [inline examples/advanced_pytorch/client.py:33]:
```suggestion
        model = load_model() # Same function that creates the model. 
        model.load_state_dict(state_dict, strict=True)
        return model
```

File: examples/advanced_pytorch/client.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+        valset = torch.utils.data.Subset(self.trainset, range(0, n_valset))
+        trainset = torch.utils.data.Subset(
+            self.trainset, range(n_valset, len(self.trainset))
+        )
+
+        trainLoader = DataLoader(trainset, batch_size=batch_size, shuffle=True)
+        valLoader = DataLoader(valset, batch_size=batch_size)
+
+        results = utils.train(self.model, trainLoader, valLoader, epochs)
```
Comment (pedropgusmao) [inline examples/advanced_pytorch/client.py:55]: Removing `self`
```suggestion
        results = utils.train(model, trainLoader, valLoader, epochs)
```

Comment (pedropgusmao) [inline examples/advanced_pytorch/client.py:55]: Removing `self`
```suggestion
        results = utils.train(self.model, trainLoader, valLoader, epochs)
```

File: examples/advanced_pytorch/client.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+            val.cpu().numpy() for _, val in self.model.state_dict().items()
+        ]
+        num_examples_train = len(trainset)
+
+        return parameters_prime, num_examples_train, results
+
+    def evaluate(self, parameters, config):
+        """Evaluate parameters on the locally held test set."""
+        # Update local model parameters
+        self.set_parameters(parameters)
```
Comment (pedropgusmao) [inline examples/advanced_pytorch/client.py:67]: Using returned model
```suggestion
        model = self.set_parameters(parameters)
```

File: examples/advanced_pytorch/client.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+        # Get config values
+        steps: int = config["val_steps"]
+
+        # Evaluate global model parameters on the local test data and return results
+        testloader = DataLoader(self.testset, batch_size=16)
+
+        loss, accuracy = utils.test(self.model, testloader, steps)
```
Comment (pedropgusmao) [inline examples/advanced_pytorch/client.py:75]: Removing `self`
```suggestion
        loss, accuracy = utils.test(model, testloader, steps)
```

File: examples/advanced_pytorch/client.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+        return float(loss), len(self.testset), {"accuracy": float(accuracy)}
+
+
+def client_dry_run():
+    """Weak tests to check whether all client methods are working as expected."""
+    model = utils.load_efficientnet(classes=10)
+    trainset, testset = utils.load_partition(0)
+    trainset = torch.utils.data.Subset(trainset, range(10))
+    testset = torch.utils.data.Subset(testset, range(10))
+    client = CifarClient(model, trainset, testset)
```
Comment (pedropgusmao) [inline examples/advanced_pytorch/client.py:85]: No need to pass a model anymore.
```suggestion
    client = CifarClient(trainset, testset)
```

File: examples/advanced_pytorch/client.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+    # Load a subset of CIFAR-10 to simulate the local data partition
+    trainset, testset = utils.load_partition(args.partition)
+
+    if args.toy:
+        trainset = torch.utils.data.Subset(trainset, range(10))
+        testset = torch.utils.data.Subset(testset, range(10))
+
+    # Start Flower client
+    client = CifarClient(model, trainset, testset)
```
Comment (pedropgusmao) [inline examples/advanced_pytorch/client.py:124]: No need to pass a model anymore
```suggestion
    client = CifarClient(trainset, testset)
```

File: examples/advanced_pytorch/client.py [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+
+def client_dry_run():
+    """Weak tests to check whether all client methods are working as expected."""
+    model = utils.load_efficientnet(classes=10)
+    trainset, testset = utils.load_partition(0)
+    trainset = torch.utils.data.Subset(trainset, range(10))
+    testset = torch.utils.data.Subset(testset, range(10))
+    client = CifarClient(model, trainset, testset)
+    client.fit(
+        [val.cpu().numpy() for _, val in model.state_dict().items()],
```
Comment (pedropgusmao) [inline examples/advanced_pytorch/client.py:87]: This looks a bit obscure... Should this become a function like `load_weights(model)` or something?  
```suggestion
        [val.cpu().numpy() for _, val in model.state_dict().items()],
```

File: examples/advanced_pytorch/client.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+        default=False,
+        required=False,
+        help="Set to true to quicky run the client using only 10 datasamples. \
+        Useful for testing purposes. Default: False",
+    )
+
+    args = parser.parse_args()
+
+    # Load PyTorch model
+    model = utils.load_efficientnet(classes=10)
```
Comment (pedropgusmao) [inline examples/advanced_pytorch/client.py:114]: No need to create the model anymore
```suggestion

```

File: examples/advanced_pytorch/utils.py [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+        net.parameters(), lr=0.1, momentum=0.9, weight_decay=1e-4
+    )
+    net.train()
+    for _ in range(epochs):
+        for images, labels in trainloader:
+            images, labels = images.to(DEVICE), labels.to(DEVICE)
+            optimizer.zero_grad()
+            loss = criterion(net(images), labels)
+            loss.backward()
+            optimizer.step()
```
Comment (pedropgusmao) [inline examples/advanced_pytorch/utils.py:61]: Making sure `train` receives a device. 
```suggestion
def train(net, trainloader, valloader, epochs, device= "cpu"):
    """Train the network on the training set."""
    print("Starting training...")
    net.to(device)  # move model to GPU if available
    criterion = torch.nn.CrossEntropyLoss().to(device)
    optimizer = torch.optim.SGD(
        net.parameters(), lr=0.1, momentum=0.9, weight_decay=1e-4
    )
    net.train()
    for _ in range(epochs):
        for images, labels in trainloader:
            images, labels = images.to(device), labels.to(device)
            optimizer.zero_grad()
            loss = criterion(net(images), labels)
            loss.backward()
            optimizer.step()
```

File: examples/advanced_pytorch/utils.py [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+def test(net, testloader, steps: int = None):
+    """Validate the network on the entire test set."""
+    print("Starting evalutation...")
+    net.to(DEVICE)  # move model to GPU if available
+    criterion = torch.nn.CrossEntropyLoss()
+    correct, total, loss = 0, 0, 0.0
+    net.eval()
+    with torch.no_grad():
+        for batch_idx, (images, labels) in enumerate(testloader):
+            images, labels = images.to(DEVICE), labels.to(DEVICE)
```
Comment (pedropgusmao) [inline examples/advanced_pytorch/utils.py:86]:
```suggestion
def test(net, testloader, device = "cpu", steps: int = None):
    """Validate the network on the entire test set."""
    print("Starting evalutation...")
    net.to(device)  # move model to GPU if available
    criterion = torch.nn.CrossEntropyLoss()
    correct, total, loss = 0, 0, 0.0
    net.eval()
    with torch.no_grad():
        for batch_idx, (images, labels) in enumerate(testloader):
            images, labels = images.to(device), labels.to(device)
```

File: examples/advanced_pytorch/server.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+    # Parse command line argument `partition`
+    parser = argparse.ArgumentParser(description="Flower")
+    parser.add_argument(
+        "--toy",
+        type=bool,
+        default=False,
+        required=False,
+        help="Set to true to use only 10 datasamples for validation. \
+            Useful for testing purposes. Default: False",
+    )
```
Comment (pedropgusmao) [inline examples/advanced_pytorch/server.py:84]: parser.add_argument(
        "--use_cuda",
        type=bool,
        default=False,
        required=False,
        help="Set to true to use GPU. Default: False",
    )

File: examples/advanced_pytorch/server.py [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+
```
Comment (pedropgusmao) [inline examples/advanced_pytorch/server.py:85]: Allow user to decide whether or not to use GPU (cuda). 
```suggestion
    parser.add_argument(
        "--use_cuda",
        type=bool,
        default=False,
        required=False,
        help="Set to true to use GPU. Default: False",
    )
```

File: examples/advanced_pytorch/client.py [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+    parser.add_argument("--partition", type=int, choices=range(0, 10), required=True)
+    parser.add_argument(
+        "--toy",
+        type=bool,
+        default=False,
+        required=False,
+        help="Set to true to quicky run the client using only 10 datasamples. \
+        Useful for testing purposes. Default: False",
+    )
+
```
Comment (pedropgusmao) [inline examples/advanced_pytorch/client.py:110]: Allow user to decide whether or not to use GPU (cuda). 
```suggestion
    parser.add_argument(
        "--use_cuda",
        type=bool,
        default=False,
        required=False,
        help="Set to true to use GPU. Default: False",
    )
```

File: examples/advanced_pytorch/client.py [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+warnings.filterwarnings("ignore")
+
+
+class CifarClient(fl.client.NumPyClient):
+    def __init__(
+        self,
+        model: torch.nn.Module,
+        trainset: torchvision.datasets,
+        testset: torchvision.datasets,
+        validation_split: int = 0.1,
```
Comment (pedropgusmao) [inline examples/advanced_pytorch/client.py:19]: Include `device` as parameter.
```suggestion
        device: str = "cpu",
        validation_split: int = 0.1,
```

File: examples/advanced_pytorch/client.py [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+    ):
+        self.model = model
+        self.trainset = trainset
+        self.testset = testset
```
Comment (pedropgusmao) [inline examples/advanced_pytorch/client.py:23]:
```suggestion
        self.testset = testset
        self.device = device
```

File: examples/advanced_pytorch/client.py (hunk 4 above)
Comment (pedropgusmao) [inline examples/advanced_pytorch/client.py:55]: Pass `device` to `train`
```suggestion
        results = utils.train(self.model, trainLoader, valLoader, epochs, self.device)
```

File: examples/advanced_pytorch/client.py (hunk 6 above)
Comment (pedropgusmao) [inline examples/advanced_pytorch/client.py:75]: Pass `device` to to `test`
```suggestion
        loss, accuracy = utils.test(self.model, testloader, steps, self.device)
```

File: examples/advanced_pytorch/client.py [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+        steps: int = config["val_steps"]
+
+        # Evaluate global model parameters on the local test data and return results
+        testloader = DataLoader(self.testset, batch_size=16)
+
+        loss, accuracy = utils.test(self.model, testloader, steps)
+        return float(loss), len(self.testset), {"accuracy": float(accuracy)}
+
+
+def client_dry_run():
```
Comment (pedropgusmao) [inline examples/advanced_pytorch/client.py:79]: Add device to `dry_run`
```suggestion
def client_dry_run(device = "cpu"):
```

**Comment by pedropgusmao** [comment]
Hi @cozek,
I left a few suggestions on how to load the model only if needed and how to pass the flag whether or not to use gpus. 
Let me know what you think.

**Comment by cozek** [comment]
> Hi @cozek, I left a few suggestions on how to load the model only if needed and how to pass the flag whether or not to use gpus. Let me know what you think.
Thanks for the suggestions. Made the changes accordingly.

**Review by danieljanes — APPROVED** [review]

**Comment by danieljanes** [comment]
Thanks for the PR @cozek & thanks for the review @pedropgusmao 👍

**Pull request merged**